### ISSUE TITLE
Restart Policies broken on master

### ISSUE THREAD

**Issue by cpuguy83** [docker/docker#16887]
When a container restart policy kicks in, there is a networking issue when the container restart is attempted, here is the error message in the daemon logs: `ERRO[0015] Error running container: [8] System error: failed to set sandbox key : already assigned`

This happens at each restart attempt (by the restart policy).

Repro:
```
docker run -d --restart=always busybox sh -c 'exit 0'
```
Then check the daemon logs.
Note that even when the container never actually starts again due to the network error which you can also see by doing:

```
docker run -d --restart=always busybox sh -c "trap 'exit 0' TERM; while true; do sleep 1; done"
```
And then killing the container pid.

ping @mavenugo

**Comment by jfrazelle** [comment]
how do we not have an integration test for this :(

**Comment by cpuguy83** [comment]
Also a note for testing this... the `waitRun` call in `integration-cli/docker_utils.go` waits for `containter.State.Running == true`... which is true even when the actual state is restarting.
So have to look at `container.State.Status`

Here's a diff with a test for this:
```diff
diff --git a/integration-cli/docker_cli_restart_test.go b/integration-cli/docker_cli_restart_test.go
index da38056..4571a17 100644
--- a/integration-cli/docker_cli_restart_test.go
+++ b/integration-cli/docker_cli_restart_test.go
@@ -148,3 +148,22 @@ func (s *DockerSuite) TestContainerRestartwithGoodContainer(c *check.C) {
 	}
 
 }
+
+func (s *DockerSuite) TestContainerRestartSuccess(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	out, _ := dockerCmd(c, "run", "-d", "--restart=always", "busybox", "top")
+	id := strings.TrimSpace(out)
+	c.Assert(waitRun(id), check.IsNil)
+
+	pid, err := inspectField(id, "State.Pid")
+	c.Assert(err, check.IsNil)
+	dockerCmd(c, "run", "--pid=host", "busybox", "kill", "-9", pid)
+
+	count, err := inspectField(id, "RestartCount")
+	c.Assert(err, check.IsNil)
+	c.Assert(count, check.Equals, "1")
+
+	err = waitInspect(id, "{{.State.Status}}", "running", 5*time.Second)
+	c.Assert(err, check.IsNil)
+}
```

**Comment by mavenugo** [comment]
Thanks @cpuguy83 for the tests. We can push it as a separate PR.

But the issue itself is fixed by #16910. lets close it.
